### ISSUE TITLE
cannot create hardlinks

### ISSUE THREAD

**Issue by HOSTED-POWER** [mysecureshell/mysecureshell#52]
Hi,

Hard links are used by many backup programs and are really useful, but using mysecureshell I can no longer create hard links. 

When disabling mysecureshell and using the regular shells it works perfectly.

root@web [/backup]# ln foo bar
ln: failed to create hard link ‘bar’ => ‘foo’: Function not implemented

Is there some setting to adjust? I already played with ShowLinksAsLinks but no luck.

Thx a lot for fixing this!

**Comment by Teka101** [comment]
Hi,

SFTP protocol doesn't allow creation of hard-link. You can only do symbol-link and if you don't want to show it, you can use "ShowLinksAsLinks".

**Comment by HOSTED-POWER** [comment]
Hi Teka101,

It's already possible since 2009: https://lists.mindrot.org/pipermail/openssh-unix-dev/2009-February/027235.html

We tried with any other shell like bash and rssh and hardlinking works perfectly. Only mysecureshell is unable to support it :(

We really really need it for our backups. It would be wonderful if you could look into it!

**Comment by Teka101** [comment]
Hi,

It's a specific feature by OpenSSH, so not yet implemented in MySecureShell.

**Comment by HOSTED-POWER** [comment]
Hi,

Does this mean you're looking into it? It would be really really great!  :)

**Comment by Teka101** [comment]
I think yes 👍

**Comment by HOSTED-POWER** [comment]
Nice, any idea if this is a lot of work or any ETA? We're really in trouble with our backups. We have no clue where to look in the code either to perhaps fix it ourselves.

**Comment by Teka101** [comment]
I don't have the time to do now :/

But you can take a look at file [Sftp.c](https://github.com/mysecureshell/mysecureshell/blob/master/SftpServer/Sftp.c)  in function "DoExtended", if you want to fix it 👍

**Comment by HOSTED-POWER** [comment]
If I understood correctly I need at least sftp version 5 to support the hardlinks? (cVersion in the source)

Well I think I was close, but I'm stuck. SSHFS client insists on using sftp version 3 instead of 5, I don't understand why because with openssh only it worked to create hardlinks.

**Comment by HOSTED-POWER** [comment]
nvm it finally works!

sftp.c

```c
void DoInit()
{
	u_int32_t clientVersion;
	tBuffer *b;

	clientVersion = BufferGetInt32(bIn);
	b = BufferNew();
	if (b == NULL)
		return;
	BufferPutInt8FAST(b, SSH2_FXP_VERSION);
	connectionStatus = CONN_SFTP;
#ifdef MSS_HAVE_ADMIN
	if (clientVersion == SSH2_ADMIN_VERSION)
	{
		if (HAS_BIT(gl_var->flagsGlobals, SFTPWHO_IS_ADMIN)
				|| HAS_BIT(gl_var->flagsGlobals, SFTPWHO_IS_SIMPLE_ADMIN))
		{
			connectionStatus = CONN_ADMIN;
			cVersion = clientVersion;
			if (!HAS_BIT(gl_var->flagsGlobals, SFTPWHO_IS_ADMIN))
				cVersion = SSH2_SIMPLE_ADMIN_VERSION;
			DEBUG((MYLOG_DEBUG, "[DoInit]New admin [use version: %i]", cVersion));
			//Hide admin to sftp-who !
			gl_var->who->status = SFTPWHO_EMPTY;
			gl_var->who = NULL;
			stats = StatsNew();
			BufferPutInt32(b, cVersion);
#ifdef MSSEXT_DISKUSAGE
			BufferPutString(b, "space-available");
#endif
#ifdef MSSEXT_FILE_HASHING
			BufferPutString(b, "check-file");
#endif
		}
	}
#endif
	if (connectionStatus == CONN_SFTP)
	{
		DoInitUser();
		if (clientVersion < 3)
			cVersion = 3;
		else if (clientVersion < cVersion)
			cVersion = clientVersion;
		
		BufferPutInt32(b, cVersion);
		DEBUG((MYLOG_DEBUG, "[DoInit]New client want version: %i [use: %i]", clientVersion, cVersion));
		if (cVersion >= 4)
		{
			BufferPutString(b, "newline");
			BufferPutString(b, "\n");
			if (cVersion >= 5)
			{
				tBuffer *opt;

				BufferPutString(b, "supported");
				opt = BufferNew();
				BufferPutInt32(opt, SSH5_FILEXFER_ATTR__MASK);
				BufferPutInt32(opt, SSH5_FILEXFER_ATTR__BITS);
				BufferPutInt32(opt, SSH5_FXF__FLAGS);
				BufferPutInt32(opt, SSH5_FXF_ACCESS__FLAGS);
				BufferPutInt32(opt, SSH2_MAX_READ);
#ifdef MSSEXT_DISKUSAGE
				BufferPutString(opt, "space-available");
#endif //MSSEXT_DISKUSAGE
				BufferPutString(opt, "hardlink@openssh.com");
#ifdef MSSEXT_DISKUSAGE_SSH
				BufferPutString(opt, "statvfs@openssh.com");
				BufferPutString(opt, "fstatvfs@openssh.com");
#endif //MSSEXT_DISKUSAGE_SSH
#ifdef MSSEXT_FILE_HASHING
				BufferPutString(opt, "check-file");
#endif //MSSEXT_FILE_HASHING
				BufferPutPacket(b, opt);
				BufferDelete(opt);
			}
			else
				goto DO_EXTENSION_V3;
		}
		else
		{
			DO_EXTENSION_V3:
#ifdef MSSEXT_DISKUSAGE
			BufferPutString(b, "space-available");
			BufferPutString(b, "");
#endif //MSSEXT_DISKUSAGE
#ifdef MSSEXT_DISKUSAGE_SSH
			BufferPutString(b, "statvfs@openssh.com");
			BufferPutString(b, "2");
			BufferPutString(b, "fstatvfs@openssh.com");
			BufferPutString(b, "2");
#endif //MSSEXT_DISKUSAGE_SSH
			BufferPutString(b, "hardlink@openssh.com");
			BufferPutString(b, "1");
#ifdef MSSEXT_FILE_HASHING
			BufferPutString(b, "check-file");
			BufferPutString(b, "");
#endif //MSSEXT_FILE_HASHING
		}
	}
	BufferPutPacket(bOut, b);
	BufferDelete(b);
}

void DoExtended()
{
	u_int32_t id;
	char *request;

	id = BufferGetInt32(bIn);
	request = BufferGetString(bIn);
	DEBUG((MYLOG_DEBUG, "[DoExtended]request:'%s'", request));
#ifdef MSSEXT_DISKUSAGE
	if (strcmp(request, "space-available") == 0)
		DoExtDiskSpace(bIn, bOut, id);
	else
#endif //MSSEXT_DISKUSAGE

	if (strcmp(request, "hardlink@openssh.com") == 0)
		DoExtHardLink(bIn, bOut, id);
	else

#ifdef MSSEXT_DISKUSAGE_SSH
	if (strcmp(request, "statvfs@openssh.com") == 0)
		DoExtDiskSpaceOpenSSH_Name(bIn, bOut, id);
	else if (strcmp(request, "fstatvfs@openssh.com") == 0)
		DoExtDiskSpaceOpenSSH_Handle(bIn, bOut, id);
	else
#endif //MSSEXT_DISKUSAGE_SSH
#ifdef MSSEXT_FILE_HASHING
	if (strcmp(request, "check-file-handle") == 0)
		DoExtFileHashing_Handle(bIn, bOut, id);
	else if (strcmp(request, "check-file-name") == 0)
		DoExtFileHashing_Name(bIn, bOut, id);
	else
#endif //MSSEXT_FILE_HASHING
		SendStatus(bOut, id, SSH2_FX_OP_UNSUPPORTED);
	free(request);
}

```
sftpext.c

```c

void DoExtHardLink(tBuffer *bIn, tBuffer *bOut, u_int32_t id)
{
	char *oldPath, *newPath;
	int r, status;

	oldPath = convertFromUtf8(BufferGetString(bIn), 1);
	newPath = convertFromUtf8(BufferGetString(bIn), 1);

	DEBUG((MYLOG_DEBUG, "[DoExtHardLink]oldPath: %s, newPath: %s", oldPath, newPath));

	r = link(oldPath, newPath);
	status = (r == -1) ? errnoToPortable(errno) : SSH2_FX_OK;
	SendStatus(bOut, id, status);

	free(oldPath);
	free(newPath);
}
```

Now the code it pretty clean, but it will hardcode it into the source. No option to disable/enable via configure etc...

 How could we proceed to get it into the mysecureshell? It's been quite some work for me already to get this working since I don't have much experience with this particular source.

It would be great if you could merge the changes somehow in the correct way.

**Comment by HOSTED-POWER** [comment]
PS: I cleaned up a little bit and created a patch.

**Comment by Teka101** [comment]
Fix with PR #84
